### ISSUE TITLE
Forum and nginx crashing kubernetes

### ISSUE THREAD

**Issue by DiegoGallegos4** [regisb/tutor#187]
## Steps to reproduce
Run 
```
minikube start
tutor k8s quickstart
tutor k8s databases
```
## Unexpected behavior
Forum and Nginx fails to start 

`kubectl get pods -n openedx`

```
NAME                            READY   STATUS             RESTARTS   AGE
cms-775b8bb5d8-hcvh9            1/1     Running            0          39m
elasticsearch-8df688cc7-mjj7v   1/1     Running            0          39m
forum-59f8c9f768-nzfsh          0/1     CrashLoopBackOff   10         39m
lms-77866f6c9d-gv2pn            1/1     Running            0          39m
memcached-ffc6dd968-7mxtl       1/1     Running            0          39m
mongodb-6ddbd56dcb-xdk86        1/1     Running            0          39m
mysql-644c98496c-7bnkm          1/1     Running            1          39m
nginx-6b677d6589-kpn9j          0/1     CrashLoopBackOff   11         39m
rabbitmq-6cfb745f64-p6j99       1/1     Running            0          39m
```

**Forum logs:** 
```
>>kubectl logs forum-59f8c9f768-nzfsh -n openedx
W, [2019-03-27T18:03:33.516698 #7]  WARN -- : NewRelic agent library not installed
/openedx/cs_comments_service/vendor/bundle/ruby/2.4.0/gems/elasticsearch-transport-1.1.2/lib/elasticsearch/transport/transport/base.rb:52: warning: constant ::Fixnum is deprecated
/openedx/cs_comments_service/vendor/bundle/ruby/2.4.0/gems/elasticsearch-transport-1.1.2/lib/elasticsearch/transport/transport/base.rb:54: warning: constant ::Fixnum is deprecated
W, [2019-03-27T18:03:33.804034 #7]  WARN -- : Overwriting existing field _id in class User.
W, [2019-03-27T18:03:33.837067 #7]  WARN -- : MONGODB | Unsupported client option 'max_retries'. It will be ignored.
W, [2019-03-27T18:03:33.837318 #7]  WARN -- : MONGODB | Unsupported client option 'retry_interval'. It will be ignored.
W, [2019-03-27T18:03:33.837474 #7]  WARN -- : MONGODB | Unsupported client option 'timeout'. It will be ignored.
W, [2019-03-27T18:03:34.076647 #7]  WARN -- : NewRelic agent library not installed
ERROR: ElasticSearch configuration validation failed. "rake search:validate_index" failed with the following message: Alias 'content' does not exist.
```

**Nginx logs**
```
>> kubectl logs nginx-6b677d6589-kpn9j -n openedx
2019/03/27 18:02:52 [emerg] 1#1: host not found in upstream "notes:8000" in /etc/nginx/conf.d/extra.conf:3
nginx: [emerg] host not found in upstream "notes:8000" in /etc/nginx/conf.d/extra.conf:3
```
## Additional info (IMPORTANT)

`tutor, version 3.1.0`

Thus making migrations to fail: 

```
Running forum migrations...
kubectl exec --namespace openedx -it forum-59f8c9f768-nzfsh -- bash -e -c bundle exec rake search:initialize && bundle exec rake search:rebuild_index
error: unable to upgrade connection: container not found ("forum")
Running notes migrations...
Error: Pod with app {} does not exist. Make sure that the pod is running.
```

**Comment by regisb** [comment]
Hi @DiegoGallegos4! Thanks for opening this issue. So this is kind of a double issue: there is the notes issue, and the forum issue. The forum issue is actually related to #179 -- or so I believe. Basically, unless I'm mistaken, the forum takes some time to start, so the quickstart fails the first time you run it.

As for the notes issue, that is indeed a new bug. Can you please confirm that the quickstart succeeds after a couple tries if you disable notes?

**Comment by DiegoGallegos4** [comment]
`tutor k8s quickstart`

Does succeed:
```
2019/03/27 18:31:55 Problem with dial: dial tcp 10.104.252.34:3306: i/o timeout. Sleeping 1s
2019/03/27 18:31:55 Timeout after 20s waiting on dependencies to become available: [tcp://mysql:3306]
command terminated with exit code 1
Running forum migrations...
kubectl exec --namespace openedx -it forum-59f8c9f768-nzfsh -- bash -e -c bundle exec rake search:initialize && bundle exec rake search:rebuild_index
error: unable to upgrade connection: container not found ("forum")
Creating oauth2 users...
kubectl exec --namespace openedx -it lms-77866f6c9d-gv2pn -- bash -e -c ./manage.py lms create_oauth2_client     "http://androidapp.com" "http://androidapp.com/redirect" public     --client_id android --client_secret MMAWj5oG6AdwzmN3jGQv3lwu     --trusted
Databases ready.
```

Forum still on **CrashLoop**: (no notes)
```
NAME                            READY   STATUS             RESTARTS   AGE
cms-775b8bb5d8-hcvh9            1/1     Running            0          75m
elasticsearch-8df688cc7-mjj7v   1/1     Running            0          75m
forum-59f8c9f768-nzfsh          0/1     CrashLoopBackOff   16         75m
lms-77866f6c9d-gv2pn            1/1     Running            0          75m
memcached-ffc6dd968-7mxtl       1/1     Running            0          75m
mongodb-6ddbd56dcb-xdk86        1/1     Running            0          75m
mysql-644c98496c-7bnkm          1/1     Running            2          75m
nginx-6b677d6589-kpn9j          1/1     Running            17         75m
rabbitmq-6cfb745f64-p6j99       1/1     Running            0          75m
```

**Comment by regisb** [comment]
Can you get some logs from the forums to understand why they are crashing?

**Comment by DiegoGallegos4** [comment]
Logs: 
```
>>> (master) tutor $ kubectl logs forum-6f7668ddb5-54tmx -n openedx
W, [2019-03-28T02:06:15.202691 #8]  WARN -- : NewRelic agent library not installed
/openedx/cs_comments_service/vendor/bundle/ruby/2.4.0/gems/elasticsearch-transport-1.1.2/lib/elasticsearch/transport/transport/base.rb:52: warning: constant ::Fixnum is deprecated
/openedx/cs_comments_service/vendor/bundle/ruby/2.4.0/gems/elasticsearch-transport-1.1.2/lib/elasticsearch/transport/transport/base.rb:54: warning: constant ::Fixnum is deprecated
W, [2019-03-28T02:06:16.111033 #8]  WARN -- : Overwriting existing field _id in class User.
W, [2019-03-28T02:06:16.462478 #8]  WARN -- : MONGODB | Unsupported client option 'max_retries'. It will be ignored.
W, [2019-03-28T02:06:16.462677 #8]  WARN -- : MONGODB | Unsupported client option 'retry_interval'. It will be ignored.
W, [2019-03-28T02:06:16.462705 #8]  WARN -- : MONGODB | Unsupported client option 'timeout'. It will be ignored.
W, [2019-03-28T02:06:16.470964 #8]  WARN -- : NewRelic agent library not installed
ERROR: ElasticSearch configuration validation failed. "rake search:validate_index" failed with the following message: Alias 'content' does not exist.
```

**Comment by regisb** [comment]
@DiegoGallegos4 I need to sit down and actually launch a real k8s cluster to debug this one. Unfortunately I won't be able to do that before two weeks, sorry :-/ Is that an urgent issue?

**Comment by DiegoGallegos4** [comment]
no not urgent, I've been trying running to run tutor k8s for a while now still not luck.

**Comment by arbrandes** [comment]
Just a quick note to say I am able to reproduce the forum issue exactly.  Running on a microk8s cluster, but I doubt this makes a difference.

```
/openedx/cs_comments_service/vendor/bundle/ruby/2.4.0/gems/elasticsearch-transport-1.1.2/lib/elasticsearch/transport/transport/base.rb:52: warning: constant ::Fixnum is deprecated
/openedx/cs_comments_service/vendor/bundle/ruby/2.4.0/gems/elasticsearch-transport-1.1.2/lib/elasticsearch/transport/transport/base.rb:54: warning: constant ::Fixnum is deprecated
W, [2019-04-02T12:55:56.083848 #8]  WARN -- : Overwriting existing field _id in class User.
W, [2019-04-02T12:55:56.163861 #8]  WARN -- : MONGODB | Unsupported client option 'max_retries'. It will be ignored.
W, [2019-04-02T12:55:56.163965 #8]  WARN -- : MONGODB | Unsupported client option 'retry_interval'. It will be ignored.
W, [2019-04-02T12:55:56.164007 #8]  WARN -- : MONGODB | Unsupported client option 'timeout'. It will be ignored.
ERROR: ElasticSearch configuration validation failed. "rake search:validate_index" failed with the following message: Alias 'content' does not exist.
```

The same forum image otherwise works on docker-compose.

**Comment by DiegoGallegos4** [comment]
@regisb where you able to check this?

**Comment by regisb** [comment]
@DiegoGallegos4 Yes, absolutely. The forum issue is due to the fact that the forum can't boot if Elasticsearch is not configured. But we can't run migrations and configure Elasticsearch if the forum container is not running, in the current state.
The solution to this egg-and-chicken problem will consist in writing the database migration task as a k8s job, instead of doing a `kubectl exec`.
Getting k8s to work properly is the first item on my todo list. You can expect the k8s integration to come out of beta in the next couple days/weeks.

**Comment by DiegoGallegos4** [comment]
Ok cool, if there is anything I can help you with it'll be great. Not very experienced with k8s but I am an experienced dev and if you got a spec to follow I can help you out. Great work!

**Comment by yamashi** [comment]
My temporary single line work around: 

``kubectl run es-deploy --rm -i --tty --image regis/openedx-forum:ironwood -n openedx -- bash -c "bundle exec rake search:initialize && bundle exec rake search:rebuild_index"``

**Comment by regisb** [comment]
This should be solved in v3.4.0.
